Feat/add binary comparison

Michal-MK opened the pull request:
- added net7.0 target
- added CheckBinary function to well... check binary content
- added GetOldBinaryContent which avoids the terminal and string.Join("\n")
- added CheckOutputBinaryCore which avoids string.Replace and other undesirable operations on binary files

Michal-MK committed:
Add binary comparison option
- added net7.0 target
- added CheckBinary function to well... check binary content
- added GetOldBinaryContent which avoids the terminal and string.Join("\n")
- added CheckOutputBinaryCore which avoids string.Replace and other undesirable operations on binary files

Michal-MK committed:
Cleanup

exyi commented on src/OutputChecker.cs:
```diff
@@ -160,6 +160,11 @@ private string GetOldContent(string file)
 
+        private byte[] GetOldBinaryContent(string file)
+        {
+            return File.ReadAllBytes(file);
```
This has to take the staged value from git (same as the `GetOldContent` function does it). Otherwise, on the first run the test will fail, but overwrite the file with the new (probably broken) value. Next run will read the already broken file, and pass. If we read the file from git stage, the user has to explicitly use `git add` to overwrite the file, so just running the tests mutliple times won't fix the test failure.

exyi replied:
`git cat-file` should just return the binary data to stdout, these low level git commands usually don't care about text/binary distinctions

exyi reviewed (commented):
Cool, thank you, this feature makes sense to me. I think I found a mistake GetOldBinaryContent, other than that it's ok.

Could you maybe add a unit test demonstrating that it works? (in the `exampleTest` folder, nothing complex, just `new byte[] { 0, 1, 2 }` or something)

Michal-MK committed:
Add proper handling of git-tracked files
- added BinaryTests to varify functionality
- added test Image.png as the base file and respective test outputs for each function

Michal-MK commented:
I had to restructure the code a bit as `StreamReader.ReadLine` will not work well for binary files (stripping '\n', '\r' and '\r\n' and now which one was it...?). There is a minor quirk with the `CheckModifiedBinaryData` test as it modifies the test file with the `CheckBinary` function call and VS is eager to reload it... so a popup has to be dismissed. But the metadata is not changed so nothing has to be staged in.

exyi commented:
Oh, I didn't anticipate .NET Process only allows reading through the StreamReader :facepalm: I think reading it using ASCII encoding into chars is also going to cause problems (like interpreting all >127 bytes as `?`), https://stackoverflow.com/questions/4143281/capturing-binary-output-from-process-standardoutput claims the only reasonable way is using StreamReader.BaseStream. Would you mind to try that out?

> There is a minor quirk with the CheckModifiedBinaryData test as it modifies the test file with the CheckBinary function call and VS is eager to reload it

I think it's because we always load the byte array full of `?`, so it thinks the file has changed and always overwrites it. Normally, the code avoids writing into the file if it didn't change.

Michal-MK committed:
Yep, ASCII has its own problems indeed.
- added more tests to check characters converted to bytes with different encoding

Michal-MK commented:
Right, I reverted the encoding change and added two more tests to check how it will behave with UTF-8 and ASCII. Seems good from the very limited test suite ;) Also the popup went away after restarting VS, so I do not know if/when it will come back.

Michal-MK committed:
Cleanup and Fix file not reverted to the original state

Michal-MK commented:
The popup only shows when I have the file in question open in VS. So that is not an issue as it would popup regardless. I also forgot to revert the file back to the original content so that is fixed now.

exyi commented:
Great, thank you!

I'm just wondering why I can't run github Actions on the PR. I will "solve" the issue by merging it, but if you'd know by chance, let me know.

> The popup only shows when I have the file in question open in VS

Ok then, I guess that's expected when the file changes.

> Seems good from the very limited test suite ;)

I'm afraid you added more tests than I had for the string comparisons :grin:
